### ISSUE TITLE
Cursor pixmap behaviour unstable after 27 march on 1.10.1-SNAPSHOT

### ISSUE THREAD

**Issue by redy5** [libgdx/libgdx#6835]
#### Issue details
Cursor started behaving very unstable, its pixmap has a lot of unnecessary artefact pixels and is cluttered with other pixmaps loaded before.

#### Reproduction steps/code
This is the code that changes cursors. I have experimented, changed to directly loading file to pixmap but the issue still persists. Attaching a test project I did that still has issue, also tested on other PC and issue still there. Java version 1.8.0_321.

`
int index = 0;

            for (int i = 0; i < 3; i++)

                for (int j = 0; j < 3; j++) {

                    TextureRegion[][] cursor_sprites_split = TextureRegion.split(new Texture("images/cursor_big.png"), 64, 64);

                    TextureRegion textureRegion = cursor_sprites_split[i][j];

                    TextureData textureData = textureRegion.getTexture().getTextureData();

                    if (!textureData.isPrepared())

                        textureData.prepare();

                    Pixmap pixmap = new Pixmap(textureRegion.getRegionWidth(), textureRegion.getRegionHeight(), textureData.getFormat());

                    pixmap.drawPixmap(textureData.consumePixmap(), 0, 0, textureRegion.getRegionX(), textureRegion.getRegionY(), textureRegion.getRegionWidth(), textureRegion.getRegionHeight());

                    cursors[index++] = Gdx.graphics.newCursor(pixmap, 0, 0);

                    pixmap.dispose();

                }`

#### Version of libGDX and/or relevant dependencies
        gdxVersion = '1.10.1-SNAPSHOT'
        roboVMVersion = '2.3.15'
        box2DLightsVersion = '1.5'
        ashleyVersion = '1.7.4'
        aiVersion = '1.8.2'
        gdxControllersVersion = '2.2.1'

#### Please select the affected platforms
- [ ] Android
- [ ] iOS
- [ ] HTML/GWT
- [x ] Windows
- [ ] Linux
- [ ] macOS
- [ ] 
[test.zip](https://github.com/libgdx/libgdx/files/8357450/test.zip)

Update 28.03.2022: after reverting [this](https://github.com/libgdx/libgdx/pull/6836) everything went back to normal.

**Comment by yuripourre** [comment]
@redy5 Thank you so much for reporting. I am working on it.

**Comment by tommyettinger** [comment]
This is strange...  I built and ran the test.zip (the one posted to Discord, not sure if it is different from the one posted here), and it produces what seems to be correct output.

![https://i.imgur.com/zNXrB0E.png](https://i.imgur.com/zNXrB0E.png)

I'll try with a JDK 8 version. The test.zip I used initially had this line in gradle.properties:

`org.gradle.java.home=C\:\\Users\\admin\\.jdks\\openjdk-18`

OpenJDK 18 isn't compatible with current Gradle releases, and I wonder if that might be related somehow.

**Comment by yuripourre** [comment]
@redy5 I could run your project and couldn't replicate what you are describing on a Linux environment. 
@tommyettinger thank you for looking into this too, I had the same outcome (everything is working) on Java 14.

From the logs attached I could found this, that seems related to the `drawPixmap` call and not the `decodeImage` (method changed by the commit):

```
Current thread (0x0000019798ccaff0):  JavaThread "main" [_thread_in_native, id=3564, stack(0x0000000d91f00000,0x0000000d92000000)]

Stack: [0x0000000d91f00000,0x0000000d92000000],  sp=0x0000000d91ffed48,  free space=1019k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [gdx64.dll+0x5410]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.badlogic.gdx.graphics.g2d.Gdx2DPixmap.drawPixmap(JJIIIIIIII)V+0
j  com.badlogic.gdx.graphics.g2d.Gdx2DPixmap.drawPixmap(Lcom/badlogic/gdx/graphics/g2d/Gdx2DPixmap;IIIIII)V+22
j  com.badlogic.gdx.graphics.Pixmap.drawPixmap(Lcom/badlogic/gdx/graphics/Pixmap;IIIIII)V+18
j  com.badlogic.gdx.graphics.Pixmap.drawPixmap(Lcom/badlogic/gdx/graphics/Pixmap;II)V+14
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Cursor.<init>(Lcom/badlogic/gdx/backends/lwjgl3/Lwjgl3Window;Lcom/badlogic/gdx/graphics/Pixmap;II)V+281
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Graphics.newCursor(Lcom/badlogic/gdx/graphics/Pixmap;II)Lcom/badlogic/gdx/graphics/Cursor;+11
j  com.mygdx.game.MyGdxGame.create()V+90
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.initializeListener()V+11
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update()Z+8
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop()V+111
j  com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lcom/badlogic/gdx/ApplicationListener;Lcom/badlogic/gdx/backends/lwjgl3/Lwjgl3ApplicationConfiguration;)V+269
j  com.mygdx.game.DesktopLauncher.main([Ljava/lang/String;)V+26
v  ~StubRoutines::call_stub

siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0x00000197caa700f0
```

**Comment by redy5** [comment]
@tommyettinger Copying the test.zip form discord
[test.zip](https://github.com/libgdx/libgdx/files/8359867/test.zip)
It contains jar in desktop/build/libs I built on my machine, and I run this jar on different machine to confirm its not a hardware problem, and it replicates the issue, so perhaps this is really issue with my setup, I will experiment with other JDK and update this thread. 
But it was working perfectly fine before march 27 on 1.10.1-SNAPSHOT and it works that way when I revert to 1.10.0, so something is definetely conflicting. I will make screenshots with capture tool later and update on how it looks.

**Comment by tommyettinger** [comment]
I tried the Discord test.zip with Java 8 (AdoptOpenJDK) and Java 17 (Adoptium). I can try the zip here as well, if it is different.

**Comment by tommyettinger** [comment]
There actually is an issue with the JAR in your test.zip, though the issue doesn't appear on what I built... 

![https://i.imgur.com/qro1AiE.png](https://i.imgur.com/qro1AiE.png)

In the eye at the right of the image, there's two "mystery pixels" that show up when I run the JAR. They don't appear when I compile it myself...

**Comment by tommyettinger** [comment]
Comparing the JARs, I find that the DLLs have different sizes.

![](https://i.imgur.com/NElZvrQ.png)

This suggests we built with different snapshot versions. It also might mean (if I'm actually on the latest snapshot, I might not be) that the issue showed up earlier and was fixed earlier? I'll look into this more.

**Comment by tommyettinger** [comment]
OK, NOW I can reproduce the bug. I must have been using mavenLocal()'s self-installed version this whole time. The bug looks like this, very strange:

![](https://i.imgur.com/iC5J3Nq.png)

**Comment by tommyettinger** [comment]
This gets even stranger... I haven't yet pulled the changes that were just made to revert stb_image, so I still am using the newer stb_image that might have threading bugs. I self-installed this version of libGDX (using a commit by NathanSweet from 13 hours ago) and made sure I didn't depend on an online -SNAPSHOT version. Now it seems to work... This is very strange indeed. This zip also includes a JAR built in desktop/build/libs/ , so you can compare it. If this is a threading issue, it might occur infrequently...
[test.zip](https://github.com/libgdx/libgdx/files/8360579/test.zip)

**Comment by redy5** [comment]
@tommyettinger last pic is exactly how it looks for me. In my project I cycle 9 times and create 9 different cursors with different images to store for later, and in runtime I switch between them depending on player actions, and how it actually looks:
![20220327_123032](https://user-images.githubusercontent.com/25421858/160359875-a1018e39-1c57-4358-9aa5-36fb815f8a38.jpg)
![20220327_122753](https://user-images.githubusercontent.com/25421858/160359887-5040b014-372b-4d33-a672-0765b5f09780.jpg)
Screenshots are taken from discord discussion we had yesterday since now it works and I cant reproduce.
First screenshot is when I load texture once, outside the loop, and then consume pixmaps from texture regions obtained through split. Second screenshot is when I load the texture every iteration of the loop. It was same for when I loaded image directly through new Pixmap() and then drawing fragments of it to smaller pixmaps.

In order they are loaded, every next cursor bitmap contain all the previous ones underneath + those distorted pixels on top like you showed on screenshots, regardless if I dispose of pixmaps after creating cursors.

I also experimented with creating new cursor every time it is requested and here is what I found: first 10-15 calls returned cursor with distorted pixels, then there were only 2-4 mystery pixels like you showed. But when I tried to interact with UI distorted pixels started appearing at random and after a few seconds I started getting a ton of crashes from Pixmap class.

Just now I compiled my project with latest commit revert and it works fine, so the problem is there, I'm just not sure where exactly. Sadly its unreproducable right now since commit is reverted, but we have that zip with compiled jar which clearly shows the issue was there.

Sorry for lamguage, hope you can make sense of it, since english is not native to me. Have a nice day!

**Comment by tommyettinger** [comment]
It looks like there may have been issues with the clearing logic for Pixmap, but I don't think my code includes the changes that fixed it. I think we should leave this issue open until we can confirm the latest changes are finished, and that the Cursor issue is no longer a problem.

**Comment by badlogic** [comment]
Just do a ./gradlew fetchNatives and rerun the test. It will be fine.

**Comment by crykn** [comment]
@redy5 this should be fixed since #6815 and #6818 were reverted. Can you confirm this?

**Comment by tommyettinger** [comment]
All right, guess this can be closed, a month and a day later, after some truly mind-bending debugging. Hooray!
